### PR TITLE
Reintroduce the ability to configure S3 repository credentials in cluster state

### PR DESCRIPTION
**Pull request by original-brownbear**
Revert of #46147, we want to keep this functionality around for a little longer.

This change is effectively just a copy of the 7.x code that we still have for this functionality with the exception that we now log this kind of thing as a `CRITICAL` deprecation and the hack I had to add to the settings module.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/settings/SettingsModule.java [hunk 1]
```diff
@@ -210,6 +210,13 @@ private void registerSetting(Setting<?> setting) {
         }
     }
 
+    // TODO: remove this hack once we remove the deprecated ability to use repository settings in the cluster state in the S3 snapshot
+    // module
+    private boolean isS3InsecureCredentials(Setting<?> setting) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/settings/SettingsModule.java:215]: This is a weird hack admittedly, but I needed this to get the ability to register the settings for filtering again. We don't have the requirement about the name containing a dot in 7.x and since this is just temporary I figured it was good enough.

File: modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/RepositoryCredentialsTests.java [hunk 2]
```diff
@@ -14,31 +14,54 @@
...
+import static org.hamcrest.Matchers.not;
 import static org.hamcrest.Matchers.notNullValue;
 
+@SuppressForbidden(reason = "test requires to set a System property to allow insecure settings when running in IDE")
 public class RepositoryCredentialsTests extends ESSingleNodeTestCase {
 
+    static {
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            // required for client settings overwriting when running in IDE
+            System.setProperty("es.allow_insecure_settings", "true");
```
Comment (rjernst) [inline modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/RepositoryCredentialsTests.java:60]: why is this needed when it is set via gradle? This doesn't seem right since it is setting the property forever, leaking into any test suites that run after it. Setting via gradle is the way to go (along with the isolated task for this test configuration).

**Review by rjernst — COMMENTED** [review]
I have a couple thoughts

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoryPlugin.java [hunk 3]
```diff
@@ -141,7 +141,9 @@ public List<Setting<?>> getSettings() {
             S3ClientSettings.USE_THROTTLE_RETRIES_SETTING,
             S3ClientSettings.USE_PATH_STYLE_ACCESS,
             S3ClientSettings.SIGNER_OVERRIDE,
-            S3ClientSettings.REGION
+            S3ClientSettings.REGION,
+            S3Repository.ACCESS_KEY_SETTING,
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoryPlugin.java:145]: We don't really have repository settings registered anywhere do we? As with my other comment, I had to add this back for the setting filtering on the REST layer to work so we don't serialize this back to the user (maybe there's a better workaround for this?).

[the remark this answers is not in the document]
Comment (rjernst) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoryPlugin.java:145]: "entity" settings as I have been calling them recently are not registered anywhere. The settings registered here in Plugin.getSettings are for node, cluster and index settings. So if the intention is to only have this settable directly in the repositories rest api, the there is no need to register it.

[the remark this answers is not in the document]
Comment (original-brownbear) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoryPlugin.java:145]: I get that we're abusing setting registering here, we definitely don't need these as cluster/node or index setting but we need the setting filtering during x-content serialization for these, which seems is only available by registering the setting as a filtered setting as done here (and in 7.x for the same reason).
Can I get the filtering in any other way?

[the remark this answers is not in the document]
Comment (rjernst) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RepositoryPlugin.java:145]: Ahh, I understand now. I don’t know of another way at the moment.

**Review by rjernst — APPROVED** [review]
Lgtm

**Comment by original-brownbear** [comment]
Thanks Ryan!
